### PR TITLE
Fix arch_build from settings

### PR DESCRIPTION
**Pull request by uilianries**
fixes https://github.com/conan-community/community/issues/194 https://github.com/conan-community/community/issues/195

### CODE REVIEW

**Comment by db4** [comment]
@uilianries I see that your PR does not actually contain `conanfile.py` changes (except throwing an exception in case of Windows/x86). How `package_id()` is expected to be fixed then?

**Comment by uilianries** [comment]
@db4 I'm forcing arch_build in build.py

Take a look on the current job: https://travis-ci.com/uilianries/conan-cmake-installer/jobs/197931253#L381

**Comment by db4** [comment]
> @db4 I'm forcing arch_build in build.py
 
but then your recipe's `package_id()` still does not depend on  `self.settings.build_arch`! How are you going to have 2 different binary packages for say Linux/x86 and Linux/x86_64?

**Comment by uilianries** [comment]
> but then your recipe's package_id() still does not depend on self.settings.build_arch! 

Why not, if I have `arch_build` listed in recipe's settings

**Comment by uilianries** [comment]
@db4 Now I see your point ... I need to remove `arch` from the recipe. But, it was an workaround to allow cross-building CMake installer.

**Comment by db4** [comment]
> @db4 Now I see your point ... I need to remove arch from the recipe

No, you need to include `arch_build` into `package_id()` (see https://docs.conan.io/en/latest/reference/conanfile/methods.html#info-discard-include-build-settings):
```pyrhon
    def package_id(self):
        self.info.include_build_settings()
        del self.info.settings.arch
        del self.info.settings.compiler
```

File: build.py [hunk 1]
```diff
@@ -21,27 +21,22 @@
...
+    if OSInfo().is_macos:
+        return "Macos"
+    if OSInfo().is_windows:
+        return "Windows"
+    return platform.system()
 
+if __name__ == "__main__":
+    arch = os.environ["CONAN_ARCHS"]
     builder = ConanMultiPackager()
+    os = detected_os()
```
Comment (SSE4) [inline build.py]: just a note - we need to write two issue, one for conan not exporting `detected_os`, and one for CPT not having environment variable for ARCH_BUILD (and OS_BUILD)

**Comment by db4** [comment]
I would also add
```python
    def build(self):
        self.settings.arch = self.settings.arch_build
```
To use correct architecture while building from sources

File: build.py (hunk 1 above)
Comment (uilianries) [reply build.py]: that's a good feature

Comment (uilianries) [reply build.py]: https://github.com/conan-io/conan/issues/5101
https://github.com/conan-io/conan-package-tools/issues/151

File: conanfile.py [hunk 2]
```diff
@@ -82,6 +81,8 @@ def config_options(self):
     def configure(self):
         if self._os == "Macos" and self._arch == "x86":
             raise ConanInvalidConfiguration("Not supported x86 for OSx")
+        elif self._os == "Windows" and self._arch == "x86":
+            raise ConanInvalidConfiguration("Not supported x86 for Windows")
```
Comment (db4) [inline conanfile.py]: Why did you add this? `arch_build==x86` is OK for Windows. I would rather add
```python
        if self.info.settings.os_build == "Windows":
            del self.info.settings.arch_build
```
to `package_id()` since win32 binary is used for any windows architecture.

**Comment by memsharded** [comment]
This issue could also be related: https://github.com/conan-io/conan/issues/5108

File: conanfile.py (hunk 2 above)
Comment (SSE4) [reply conanfile.py]: thanks, changed it

File: conanfile.py [hunk 3]
```diff
@@ -30,7 +30,10 @@ class CMakeInstallerConan(ConanFile):
     author = "Conan Community"
     homepage = "https://github.com/Kitware/CMake"
     topics = ("conan", "cmake", "build", "installer")
-    settings = "os_build", "arch_build", "compiler", "arch"
+    settings = {"os_build": ["Windows", "Linux", "Macos"],
+                "arch_build": ["x86", "x86_64"],
```
Comment (db4) [inline conanfile.py]: What is the reason to constraint `arch_build` if we allow building cmake from sources?

File: conanfile.py [hunk 4]
```diff
@@ -119,6 +120,8 @@ def build(self):
 
     def package_id(self):
         self.info.include_build_settings()
+        if self.settings.os == "Windows":
```
Comment (db4) [inline conanfile.py]: My mistake, should be `if self.settings.os_build == "Windows":`

File: conanfile.py (hunk 3 above)
Comment (SSE4) [reply conanfile.py]: good point, probably, let's allow all of them (i was thinking it's only installing archives initially)

File: conanfile.py (hunk 4 above)
Comment (SSE4) [reply conanfile.py]: fixed

**Comment by db4** [comment]
And I still insist on something like https://github.com/conan-community/conan-cmake-installer/pull/21#issuecomment-489740829
The reciope should be capable to build cmake on a consumer's host when binary package is not available and `arch!=arch_build` (and probably `os!=os_build`) The mentioned solution seems to work in my own recipes. If someone knows more elegant way, please let me know.

**Comment by SSE4** [comment]
I don't know any more elegant way, unfortunately

**Comment by uilianries** [comment]
I think it's good enough now

**Comment by uilianries** [comment]
@SSE4 I think we could adopt this recipe as the pattern for Bincrafters template. WDYT?

**Comment by db4** [comment]
> I think it's good enough now

except it fails to package cmake 2.8.12 binary distribution on Macos. The only difference from the later versions I see is a space in `CMake 2.8-12.app` directory name. Can someone with access to Macos host debug the issue?

**Comment by dragon512** [comment]
I had filed issue conan-io/conan#5108 
If I do an install via adding "--build cmake_install" the binaries works. Does this mean there is an issue with the recipe or the platform in which the binaries are built on?

**Comment by db4** [comment]
> 
> 
> I had filed issue [conan-io/conan#5108](https://github.com/conan-io/conan/issues/5108)
> If I do an install via adding "--build cmake_install" the binaries works. Does this mean there is an issue with the recipe or the platform in which the binaries are built on?

I believe it's due to currently broken cmake_installer recipe. It ignores `arch_build` and tries to use the same x86 cmake binary from `conan-center` for any Linux host architecture. Your system does not have 32-bit support installed so you get an error. Wait for this PR to be merged and things should go better.

**Comment by SSE4** [comment]
I've run local build of 2.8.12 on OSX and it passes, investigating...

**Comment by db4** [comment]
> I've run local build of 2.8.12 on OSX and it passes, investigating...

You must be trying `conan create . cmake_installer/2.8.12@conan/stable` but it's `conan create . cmake_installer/1.0@conan/stable -o cmake_installer:version=2.8.12` that fails. Now it's clear why. In https://github.com/conan-community/conan-cmake-installer/blob/b9592659973adb954823710fa7b38423328b1a40/conanfile.py#L139

`self.version` should be replaced with `self._cmake_version`.

**Comment by SSE4** [comment]
@db4 good catch! thanks a lot

**Comment by db4** [comment]
@SSE4 Is it time to merge this PR?

**Comment by SSE4** [comment]
@db4 indeed, it has finished finally!

**Comment by db4** [comment]
@SSE4 Please also promote it to `stable` channel (that was removed in https://github.com/conan-community/conan-cmake-installer/pull/21/commits/bf25e0d09cb8d5c839ff68e0d863454f1b2e0131) to update conan-center packages

**Comment by SSE4** [comment]
thanks, done
